sagemaker/pipeline/pipeline_parameters.py

```diff
@@ -0,0 +1,764 @@
+class ScriptPaths:
+    """Entry point script locations"""
+
+    dist_part_script: str
+    gb_convert_script: str
+    train_script: str
+    inference_script: str
+    gconstruct_script: str
+    gsprocessing_script: str
+
+
+@dataclass()
+class PipelineArgs:
+    """Wrapper class for all pipeline configurations.
+
+    Parameters
+    ----------
+    aws_config : AWSConfig
+        AWS configuration settings
+    graph_construction_config : GraphConstructionConfig
+        Graph construction configuration
+    instance_config : InstanceConfig
+        Instance configuration settings
+    task_config : TaskConfig
+        Task-level configuration settings
+    partition_config : PartitionConfig
+        Partition configuration settings
+    training_config : TrainingConfig
+        Training configuration settings
+    inference_config : InferenceConfig
+        Inference configuration settings
+    script_paths : ScriptPaths
+        Paths to SageMaker entry point scripts
+    step_cache_expiration : str
+        Cache expiration for pipeline steps
+    update : bool
+        Whether to update existing pipeline or create a new one
+    """
+
+    aws_config: AWSConfig
+    graph_construction_config: GraphConstructionConfig
+    instance_config: InstanceConfig
+    task_config: TaskConfig
+    partition_config: PartitionConfig
+    training_config: TrainingConfig
+    inference_config: InferenceConfig
+    script_paths: ScriptPaths
+    step_cache_expiration: str
+    update: bool
+
+    def __eq__(self, other):
+        if not isinstance(other, PipelineArgs):
+            return False
+        return asdict(self) == asdict(other)
+
+    def __hash__(self):
+        # Convert the object to a dictionary
+        obj_dict = asdict(self)
+
+        # Convert the dictionary to a JSON string
+        json_str = json.dumps(obj_dict, sort_keys=True)
+
+        # Create a hash of the JSON string
+        return hash(json_str)
+
+    def get_hash_hex(self):
+        """Get unique string hash to differentiate between pipeline settings.
+
+        We use this hash as a unique identifier of execution parameters
+        in the intermediate output paths of the pipeline. This enables
+        caching steps when the input parameters are the same for the same
+        pipeline.
+
+        Returns
+        -------
+        str
+            A hexadecimal string representation of the object's hash.
+        """
+        # Convert the object to a dictionary
+        obj_dict = asdict(self)
+        # Remove non-functional keys from the dictionary,
+        # these do not affect the outcome of the execution
+        obj_dict.pop("step_cache_expiration")
+        obj_dict.pop("update")
+
+        # Convert the dictionary to a JSON string and create an md5 hash
+        json_str = json.dumps(obj_dict, sort_keys=True)
+        return hashlib.md5(json_str.encode()).hexdigest()
+
+    def __post_init__(self):
+        if not self.instance_config.train_on_cpu:
```

Comment:
Why not move the check into instance_config's `__post_init__`?

Reply:
Good point, `assert self.instance_config.gpu_instance_type` should be moved, `assert self.aws_config.graphstorm_pytorch_gpu_image_url` needs to be here to have access to the AWS config.